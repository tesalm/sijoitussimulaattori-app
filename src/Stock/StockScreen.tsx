import React from 'react';
import { RefreshControl, ScrollView, Text } from 'react-native';
import { Card } from 'react-native-elements';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { Colors } from '../App/colors';
import { getHistory, getIntraday, getStockMetadata, refreshIntraday } from '../MarketScreen/actions';
import { Stock } from '../MarketScreen/reducer';
import { RootState } from '../redux/reducers';
import Basicinfo from './components/Basicinfo';
import Bid from './components/Bid';
import Diagram from './components/Diagram';
import { stockContainerStyles } from './styles';
import { formatRevenue } from '../util/general';
import {
  NavigationScreenProp,
  NavigationState,
  NavigationScreenProps,
} from 'react-navigation';
import BackButtonWithNavigation from '../navigation/components/BackButton';

export interface StockProps {
  getMeta: typeof getStockMetadata;
  getIntra: typeof getIntraday;
  refreshIntra: typeof refreshIntraday;
  getHistoryData: typeof getHistory;
  refreshing: boolean;
  symbol?: string;
  stock?: Stock;
}

type StockPropsWithNavigation = StockProps & NavigationScreenProps;

interface StockState {}

export class StockScreen extends React.Component<
  StockPropsWithNavigation,
  StockState
> {
  constructor(props: StockPropsWithNavigation) {
    super(props);
  }

  static navigationOptions = {
    title: 'Stock',
    headerLeft: <BackButtonWithNavigation />,
  };

  componentDidMount() {
    if (this.props.symbol && this.props.stock) {
      // Attempt to get or refresh necessary data:
      this.props.getMeta(this.props.stock, this.props.symbol);
      this.props.getHistoryData(this.props.stock, this.props.symbol);
      this.props.getIntra(this.props.stock, this.props.symbol);
    }
  }

  countRevenuePercentage() {
    if (
      this.props.stock !== undefined &&
      this.props.stock.stockInfo !== undefined &&
      this.props.stock.stockInfo.historyData !== undefined &&
      this.props.stock.stockInfo.intraday !== undefined
    ) {
      const yesterday = this.props.stock.stockInfo.historyData
        .historyDataQuote[0].close;
      const today = this.props.stock.stockInfo.intraday.intradayQuote[0].close;
      const revenue = (today - yesterday) / yesterday;
      return revenue;
    }
    return 0;
  }

  refresh = () => {
    if (this.props.symbol && this.props.stock) {
      this.props.refreshIntra(this.props.stock, this.props.symbol);
    }
  };

  render() {
    const { stock } = this.props;
    if (stock) {
      return (
        <ScrollView
          refreshControl={
            <RefreshControl
              refreshing={stock.stockInfo.refreshing}
              onRefresh={this.refresh}
              colors={[Colors.baseColor]}
            />
          }
        >
          <Card containerStyle={stockContainerStyles.basicInfo}>
            <Basicinfo
              revenue={this.countRevenuePercentage()}
              stockMetadata={stock.stockInfo.stockMetadata}
              metaLoading={stock.stockInfo.metaLoading}
              metaError={stock.stockInfo.metaError}
              intradayQuote={
                stock.stockInfo.intraday
                  ? stock.stockInfo.intraday.intradayQuote[0]
                  : undefined
              }
              fetchTime={
                stock.stockInfo.intraday
                  ? stock.stockInfo.intraday.fetchTime
                  : undefined
              }
              intraLoading={stock.stockInfo.intraLoading}
              intraError={stock.stockInfo.intraError}
              historyLoading={stock.stockInfo.historyLoading}
            />
          </Card>

          <Card containerStyle={stockContainerStyles.diagram}>
            <Diagram
              historyData={
                stock.stockInfo.historyData
                  ? stock.stockInfo.historyData.historyDataQuote
                  : []
              }
              historyLoading={stock.stockInfo.historyLoading}
              historyError={stock.stockInfo.historyError}
            />
          </Card>

          <Card containerStyle={stockContainerStyles.buttonContainer}>
            <Bid navigation={this.props.navigation} />
          </Card>
        </ScrollView>
      );
    } else {
      //TODO: Format the error message to user
      return <Text>Error, stockinfo not found! </Text>;
    }
  }
}

const mapStateToProps = (state: RootState) => ({
  symbol: state.stocksListing.currentSymbol,
  stock: state.stocksListing.stocks.find((stock) => {
    return stock.symbol === state.stocksListing.currentSymbol;
  }),
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      getMeta: getStockMetadata,
      getIntra: getIntraday,
      refreshIntra: refreshIntraday,
      getHistoryData: getHistory,
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(StockScreen);
