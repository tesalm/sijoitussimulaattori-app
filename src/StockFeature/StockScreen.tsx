import React from 'react';

import { Image, Text, View, TouchableHighlight } from 'react-native';
import { Card } from 'react-native-elements';
import { NavigationScreenProps } from 'react-navigation';

import { t } from '../assets/i18n';
import { RouteName } from '../navigation/routes';

import Basicinfo from './components/Basicinfo';
import Diagram from './components/Diagram';
import { stockStyles } from './styles';

import { getStock } from './actions'
import { RootState } from '../redux/reducers';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';

interface StockProps extends NavigationScreenProps {
  stock: Array<{
    key: string,
    bid: number,
    offer: number,
    high: number,
    low: number,
    marketValue: number,
    revenue: number}>;
  loading: boolean;
  error: Error | null;
  getSingleStock: typeof getStock;
}

class StockScreen extends React.Component<StockProps> {
  static navigationOptions = { title: t('StockPage.Title') };
  constructor(props:StockProps) {
    super(props);

    this.state = {
      stock: [],
      loading: false,
      error: null,
    }
  }

  componentDidMount() {
    this.props.getSingleStock();
  }

  render() {
    return (
      <View>
        <Card containerStyle={{ margin: 0, height: 147 }}>
          <Basicinfo></Basicinfo>
        </Card>

        <Card containerStyle={{ margin: 0, height: 200 }}>
          <Diagram></Diagram>
        </Card>

        <Card
          containerStyle={{ margin: 0, height: 50 }}>
          <TouchableHighlight
            onPress={() => this.props.navigation.navigate(RouteName.Profile)}
            style={ stockStyles.bidOfferButton }
            >
            <View style={{ flexDirection: 'row', justifyContent: 'space-between' }}>
              <Text style={ stockStyles.infoText }>
                  {t('Bid')}
              </Text>
              <Image
                source={require('../navigation/assets/close.png')}
                style={{ height: 30, width: 30 }}
              />
            </View>
          </TouchableHighlight>
        </Card>

        <Card
          containerStyle={{ margin: 0, height: 50 }}>
          <TouchableHighlight
            onPress={() => this.props.navigation.navigate(RouteName.Profile)}
            style={ stockStyles.bidOfferButton }
            >
            <View style={{ flexDirection: 'row', justifyContent: 'space-between' }}>
              <Text style={ stockStyles.infoText }>
                  {t('Offer')}
              </Text>
              <Image
                source={require('../navigation/assets/close.png')}
                style={{ height: 30, width: 30 }}
              />
            </View>
          </TouchableHighlight>
        </Card>

      </View>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  stock: state.singleStock.stock,
  loading: state.singleStock.loading,
  error: state.singleStock.error
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      getSingleStock: getStock,
    },
    dispatch
  );

  export default connect(
    mapStateToProps,
    mapDispatchToProps
  )(StockScreen);
