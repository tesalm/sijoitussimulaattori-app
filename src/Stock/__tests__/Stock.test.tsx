import * as React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import { StockScreen, StockProps } from '../StockScreen';
import {
  getMetadata,
  getIntraday,
  refreshIntraday,
} from '../../MarketScreen/actions';

describe('stock tests', () => {
  // Mock for navigation.
  const navigationMock: any = {};
  const middlewares: any = [];
  const mockStore = configureStore(middlewares);

  const defaultStockProps: StockProps = {
    stocks: [],
    getMeta: getMetadata,
    getIntra: getIntraday,
    refreshIntra: refreshIntraday,
    getHistoryData: jest.fn(),
    refreshing: false,
    symbol: undefined,
    stock: undefined,
  };

  const successStockProps: StockProps = {
    stocks: [
      {
        symbol: 'APL',
        name: 'Apple',
        high: 10.28,
        low: 9.99,
        revenue: -0.035,
        close: 14.57,
        currency: 'USD',
        stockInfo: {
          metaLoading: false,
          intraLoading: false,
          historyLoading: false,
          refreshing: false,
        },
      },
      {
        symbol: 'AAPL',
        name: 'Apple',
        high: 375.53,
        low: 354.38,
        revenue: 3894.34,
        close: 293.42,
        currency: 'USD',
        stockInfo: {
          metadata: {
            symbol: 'AAPL',
            name: 'Apple Inc.',
            type: 'string',
            region: 'string',
            marketOpen: '2018-12-12 11:20:00',
            marketClose: '2018-12-12 11:20:00',
            timeZone: 'US/Eastern',
            currency: 'USD',
            fetchTime: new Date('December 17, 2018 03:24:00'),
          },
          intraday: {
            symbol: 'AAPL',
            date: '2018-12-12',
            open: 170.4,
            high: 170.43,
            low: 165.43,
            close: 168.63,
            volume: 12279994,
            fetchTime: new Date('December 17, 2018 03:24:00'),
          },
          historydata: {
            symbol: 'AAPL',
            date: '2018-12-12',
            open: 170.4,
            high: 170.43,
            low: 165.43,
            close: 168.63,
            volume: 12279994,
            fetchTime: new Date('December 17, 2018 03:24:00'),
          },
          metaLoading: false,
          intraLoading: false,
          historyLoading: false,
          refreshing: false,
          metaError: undefined,
          intraError: undefined,
          historyError: undefined,
        },
      },
    ],
    stock: {
      symbol: 'AAPL',
      name: 'Apple',
      high: 375.53,
      low: 354.38,
      revenue: 3894.34,
      close: 293.42,
      currency: 'USD',
      stockInfo: {
        metadata: {
          symbol: 'AAPL',
          name: 'Apple Inc.',
          type: 'string',
          region: 'string',
          marketOpen: '2018-12-12 11:20:00',
          marketClose: '2018-12-12 11:20:00',
          timeZone: 'US/Eastern',
          currency: 'USD',
          fetchTime: new Date(),
        },
        intraday: {
          symbol: 'AAPL',
          date: '2018-12-12',
          open: 170.4,
          high: 170.43,
          low: 165.43,
          close: 168.63,
          volume: 12279994,
          fetchTime: new Date(),
        },
        historydata: {
          symbol: 'AAPL',
          date: '2018-12-12',
          open: 170.4,
          high: 170.43,
          low: 165.43,
          close: 168.63,
          volume: 12279994,
          fetchTime: new Date(),
        },
        metaLoading: false,
        intraLoading: false,
        historyLoading: false,
        refreshing: false,
        metaError: undefined,
        intraError: undefined,
        historyError: undefined,
      },
    },
    getMeta: jest.fn(),
    getIntra: jest.fn(),
    refreshIntra: jest.fn(),
    getHistoryData: jest.fn(),
    refreshing: false,
    symbol: 'AAPL',
  };

  it('Stockscreen renders correctly', async () => {
    const wrapper = shallow(
      <StockScreen {...defaultStockProps} {...navigationMock} />,
      { context: { store: mockStore() } }
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('Stockscreen renders correctly with stock', async () => {
    const wrapper = shallow(
      <StockScreen {...successStockProps} {...navigationMock} />,
      { context: { store: mockStore() } }
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
