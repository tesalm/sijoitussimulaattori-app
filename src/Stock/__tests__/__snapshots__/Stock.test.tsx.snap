// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stock tests Stockscreen renders correctly 1`] = `
<TouchableText
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  forwardedRef={null}
>
  Error, stockinfo not found!
</TouchableText>
`;

exports[`stock tests Stockscreen renders correctly with stock 1`] = `
<_class
  refreshControl={
    <RefreshControlMock
      colors={
        Array [
          "#004D40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
    />
  }
>
  <RefreshControlMock
    colors={
      Array [
        "#004D40",
      ]
    }
    onRefresh={[Function]}
    refreshing={false}
  />
  <View>
    <Themed.Card
      containerStyle={
        Object {
          "flex": 1,
          "margin": 0,
          "paddingBottom": 31.388235294117646,
          "paddingLeft": 34.285714285714285,
          "paddingRight": 34.285714285714285,
          "paddingTop": 31.388235294117646,
        }
      }
    >
      <Component
        fetchTime={2018-12-17T01:24:00.000Z}
        historyLoading={false}
        intraLoading={false}
        intradayQuote={
          Object {
            "close": 168.63,
            "date": "2018-12-12T00:20:00.000Z",
            "high": 170.43,
            "low": 165.43,
            "open": 170.4,
            "symbol": "AAPL",
            "volume": 12279994,
          }
        }
        metaLoading={false}
        revenue={0.03689356207341819}
        stockMetadata={
          Object {
            "currency": "USD",
            "fetchTime": 2018-12-17T01:24:00.000Z,
            "marketClose": "2018-12-12 11:20:00",
            "marketOpen": "2018-12-12 11:20:00",
            "name": "Apple Inc.",
            "region": "string",
            "symbol": "AAPL",
            "timeZone": "US/Eastern",
            "type": "string",
          }
        }
      />
    </Themed.Card>
    <Themed.Card
      containerStyle={
        Object {
          "flex": 1,
          "margin": 0,
          "paddingBottom": 31.388235294117646,
          "paddingLeft": 34.285714285714285,
          "paddingRight": 34.285714285714285,
          "paddingTop": 31.388235294117646,
        }
      }
    >
      <Diagram
        historyData={
          Array [
            Object {
              "close": 162.63,
              "date": "2018-12-15T00:00:00.000Z",
              "high": 170.43,
              "low": 165.43,
              "open": 170.4,
              "symbol": "AAPL",
              "volume": 12279994,
            },
            Object {
              "close": 168.63,
              "date": "2018-12-14T00:00:00.000Z",
              "high": 170.43,
              "low": 161.43,
              "open": 170.4,
              "symbol": "AAPL",
              "volume": 12279994,
            },
          ]
        }
        intraDay={
          Array [
            Object {
              "close": 168.63,
              "date": "2018-12-12T00:20:00.000Z",
              "high": 170.43,
              "low": 165.43,
              "open": 170.4,
              "symbol": "AAPL",
              "volume": 12279994,
            },
            Object {
              "close": 168.63,
              "date": "2018-12-12T00:10:00.000Z",
              "high": 170.43,
              "low": 165.43,
              "open": 170.4,
              "symbol": "AAPL",
              "volume": 12279994,
            },
          ]
        }
      />
    </Themed.Card>
    <Themed.Card
      containerStyle={
        Object {
          "flex": 1,
          "margin": 0,
          "paddingBottom": 31.388235294117646,
          "paddingLeft": 34.285714285714285,
          "paddingRight": 34.285714285714285,
          "paddingTop": 31.388235294117646,
        }
      }
    >
      <CardButton
        iconName="bid"
        onPress={[Function]}
        translationTitle="StockPage.Bid"
      />
    </Themed.Card>
  </View>
</_class>
`;
